### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by FenrisStargazer**
abrand3 / FenrisStargazer, no partner

### CODE REVIEW

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ public void Error_WithNullLogger_ThrowsException()
         // Arrange
 
         // Act
-        //BaseLoggerMixins.Error(null, "");
+        BaseLoggerMixins.Error(null, "");
 
         // Assert
```
Comment (alowe4) [inline Logger.Tests/BaseLoggerMixinsTests.cs:19]: Error_WithNullLogger_ThrowsException does not assert anything, 
I saw that BaseLoggerMixins can throw an ArgumentNullException
try to use Assert.ThrowsException<> to test

File: Logger/FileLogger.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    private string filePath;
+    public override string ClassName { get; set; }
+
+    public string getFilePath()
+    {
+        return filePath;
+    }
+	public FileLogger(string filePath)
+	{
+        this.filePath = filePath;
```
Comment (alowe4) [inline Logger/FileLogger.cs:16]: same as in log factory, avoid the "this" keyword, and use pascal case.

File: Logger/LogFactory.cs [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
+            if (filePath is not null)
+            {
+                FileLogger fileLogger = new FileLogger(filePath) {ClassName = className};
+                return fileLogger;
+            }
+            return null;
+        }
+        public void configureFileLogger(string filePath)
+        {
+            this.filePath = filePath;
```
Comment (alowe4) [inline Logger/LogFactory.cs:17]: 1. you want to avoid the 'this' keyword. Instead make the filePath on line 5 FilePath and use that. 
2. capitalize configure in configureFileLogger

**Review by alowe4 — CHANGES_REQUESTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:

    The current date/time ✔
    The name of the class that created the logger ✔
    The log level ✔
    The message ✔
    The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"

The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for

    Error, ✔
    Warning, ✔
    Information, and ✔
    Debug. ✔ 
Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. 
These methods should throw an exception if the BaseLogger parameter is null. ✔[not properly tested, but exists]
There are a couple example unit tests to get you started.

Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 1 above)
Comment (FenrisStargazer) [reply Logger.Tests/BaseLoggerMixinsTests.cs:19]: The assertion for that test is done through the "[ExpectedException(typeof(ArgumentNullException))]" on line 11.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 4]
```diff
@@ -26,19 +26,65 @@ public void Error_WithData_LogsMessage()
...
+        Assert.AreEqual("Message 55", logger.LoggedMessages[0].Message);
+    }
 
 }
 
+
 public class TestLogger : BaseLogger
 {
     public List<(LogLevel LogLevel, string Message)> LoggedMessages { get; } = new List<(LogLevel, string)>();
+    public override string ClassName { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
Comment (eroadhouse) [inline Logger.Tests/BaseLoggerMixinsTests.cs:87]: Here you should probably throw something else other than a NotImplementedException() since this code is technically implemented

File: Logger.Tests/FileLoggerTests.cs [hunk 5]
```diff
@@ -1,9 +1,16 @@
 ﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
 
-namespace Logger.Tests;
-
-[TestClass]
-public class FileLoggerTests
+namespace Logger.Tests
```
Comment (eroadhouse) [inline Logger.Tests/FileLoggerTests.cs:3]: Converting to file-scoped namespaces will be good here and for the future

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 6]
```diff
@@ -26,19 +26,65 @@ public void Error_WithData_LogsMessage()
...
-        //logger.Error("Message {0}", 42);
+        logger.Error("Message {0}", 42);
 
         // Assert
         Assert.AreEqual(1, logger.LoggedMessages.Count);
         Assert.AreEqual(LogLevel.Error, logger.LoggedMessages[0].LogLevel);
         Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
     }
+    [TestMethod]
+    public void Warning_LogsMessage()
```
Comment (eroadhouse) [inline Logger.Tests/BaseLoggerMixinsTests.cs:37]: I think it'd be a good idea to also test a null logger in Warning, Information, and Debug as well and not just Error

**Review by eroadhouse — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

Overall, nicely done!

File: Logger/LogFactory.cs [hunk 7]
```diff
@@ -1,10 +1,20 @@
...
+    public class LogFactory
     {
-        
-        return null;
+        private string? FilePath { get; set; }
+        public BaseLogger? CreateLogger(string className)
+        {
+            if (FilePath is not null)
+            {
+                FileLogger fileLogger = new FileLogger(FilePath) {ClassName = className};
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:10]:
```suggestion
                FileLogger fileLogger = new(FilePath) {ClassName = className};
```

File: Logger/FileLogger.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+	public FileLogger(string filePath)
+	{
+        FilePath = filePath;
+        ClassName = "";
+	}
+
+    public override void Log(LogLevel logLevel, string message)
+    {
+        StreamWriter file = new StreamWriter(FilePath, true);
+        file.WriteLineAsync(DateTime.Now + " " + ClassName + " " + logLevel + ": " + message);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:23]: I'd recommend using string interpolation here `$"{DateTime.Now} {ClassName} etc..."`

File: Logger/FileLogger.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    }
+	public FileLogger(string filePath)
+	{
+        FilePath = filePath;
+        ClassName = "";
+	}
+
+    public override void Log(LogLevel logLevel, string message)
+    {
+        StreamWriter file = new StreamWriter(FilePath, true);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:22]: Make sure you close your stream writer or warp it in a using statement.

File: Logger/LogFactory.cs [hunk 10]
```diff
@@ -1,10 +1,20 @@
...
-
-public class LogFactory
+﻿namespace Logger
 {
-    public BaseLogger CreateLogger(string className)
+    public class LogFactory
     {
-        
-        return null;
+        private string? FilePath { get; set; }
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:5]:
```suggestion
        private string? _FilePath;
```
This doesn't need to be an auto property, it can just be a private field.

File: Logger/BaseLoggerMixins.cs [hunk 11]
```diff
@@ -1,6 +1,54 @@
-﻿namespace Logger;
+﻿using System;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
-    
+    public static void Error(this BaseLogger b, string message, params object[] arguments)
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:7]:
```suggestion
    public static void Error(this BaseLogger logger, string message, params object[] arguments)
```
Avoid using single-character variable names, I'd recommend calling this something like `logger` instead.

File: Logger/BaseLoggerMixins.cs [hunk 12]
```diff
@@ -1,6 +1,54 @@
...
+    {
+        if (b is null)
+        {
+            throw new ArgumentNullException();
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:11]:
```suggestion
            throw new ArgumentNullException(nameof(logger));
```
Make sure to specify what argument is throwing the null exception

File: Logger/FileLogger.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    public override string ClassName { get; set; }
+
+    public string getFilePath()
+    {
+        return FilePath;
+    }
+	public FileLogger(string filePath)
+	{
+        FilePath = filePath;
+        ClassName = "";
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:17]:
```suggestion
```

File: Logger/BaseLogger.cs [hunk 14]
```diff
@@ -1,7 +1,8 @@
-﻿namespace Logger;
-
-public abstract class BaseLogger
+﻿namespace Logger
 {
-    public abstract void Log(LogLevel logLevel, string message);
+    public abstract class BaseLogger
+    {
+        public abstract string ClassName { get; set; }
```
Comment (mmwoodfo) [inline Logger/BaseLogger.cs:5]:
```suggestion
        public string? ClassName { get; set; }
```
This doesn't need to be abstract here, but it's not wrong depending on how you intend this to be used. Also, it's okay to mark this as nullable, you just need to make sure that you're expecting that it can be null.
